### DISCUSSION TITLE
TDDでのVlue ObejctパターンとDDDでのValu Objectについて

### DISCUSSION THREAD

**Discussion by yuriemori** [code-polaris/reading-circle-tdd#37] [General]
# TDDでのValue Object
コンストラクタで設定したインスタンス変数の値は可変であってはならない
複数インスタンス化した際に片方のインスタンスを変更したときにもう片方にも変更が及んではいけない

# DDDでのValue Object
![image](https://github.com/code-polaris/reading-circle-tdd/assets/60845198/bb154858-29ee-4bdf-ab95-4be0efbaa319)

どちらのValueObjectも特性としては「不変である」という点で同じ
https://learn.microsoft.com/ja-jp/dotnet/architecture/microservices/microservice-ddd-cqrs-patterns/implement-value-objects

`public abstract class ValueObject
{
    protected static bool EqualOperator(ValueObject left, ValueObject right)
    {
        if (ReferenceEquals(left, null) ^ ReferenceEquals(right, null))
        {
            return false;
        }
        return ReferenceEquals(left, right) || left.Equals(right);
    }

    protected static bool NotEqualOperator(ValueObject left, ValueObject right)
    {
        return !(EqualOperator(left, right));
    }

    protected abstract IEnumerable<object> GetEqualityComponents();

    public override bool Equals(object obj)
    {
        if (obj == null || obj.GetType() != GetType())
        {
            return false;
        }

        var other = (ValueObject)obj;

        return this.GetEqualityComponents().SequenceEqual(other.GetEqualityComponents());
    }

    public override int GetHashCode()
    {
        return GetEqualityComponents()
            .Select(x => x != null ? x.GetHashCode() : 0)
            .Aggregate((x, y) => x ^ y);
    }
    // Other utility methods
}`

**Comment by rinahira** [comment]
わかりやすくまとまってるブログがあったのでシェアします～:pray:
https://panda-program.com/posts/three-types-of-value-object
